File: msg-key-phrases/src/test/java/dev/number6/keyphrases/dagger/FakeComprehensionResultsModule.java
package dev.number6.keyphrases.dagger;

import dev.number6.comprehend.results.PresentableKeyPhrasesResults;
import dev.number6.message.ChannelMessages;
import dev.number6.message.ChannelMessagesToComprehensionResultsFunction;
import dev.number6.message.ComprehensionResultsConsumer;
import dagger.Module;
import dagger.Provides;

import javax.inject.Singleton;
import java.util.ArrayList;
import java.util.List;

import static org.junit.jupiter.api.Assertions.fail;

@Module
public class FakeComprehensionResultsModule {

    @Provides
    @Singleton
    public ChannelMessagesToComprehensionResultsFunction<PresentableKeyPhrasesResults> providesMessageToEntityResults() {
        return new ConfigurableResultsFunction();
    }

    @Provides
    @Singleton
    public ComprehensionResultsConsumer<PresentableKeyPhrasesResults> providesEntityResultsConsumer() {
        return new RecordingComprehensionResultsConsumer();
    }

    public static class ConfigurableResultsFunction implements ChannelMessagesToComprehensionResultsFunction<PresentableKeyPhrasesResults> {

        private PresentableKeyPhrasesResults presentableKeyPhrasesResults;

        @Override
        public PresentableKeyPhrasesResults apply(ChannelMessages channelMessages) {
            return presentableKeyPhrasesResults == null ?
                    fail("presentable key phrases results needs to be configured") :
                    presentableKeyPhrasesResults;
        }

        public void setPresentableKeyPhrasesResults(PresentableKeyPhrasesResults presentableKeyPhrasesResults) {
            this.presentableKeyPhrasesResults = presentableKeyPhrasesResults;
        }
    }

    public static class RecordingComprehensionResultsConsumer implements ComprehensionResultsConsumer<PresentableKeyPhrasesResults> {
        private List<PresentableKeyPhrasesResults> resultsConsumed = new ArrayList<>();

        public List<PresentableKeyPhrasesResults> getResultsConsumed() {
            return resultsConsumed;
        }

        @Override
        public void accept(PresentableKeyPhrasesResults presentableKeyPhrasesResults) {
            resultsConsumed.add(presentableKeyPhrasesResults);
        }
    }
}

File: msg-key-phrases/src/test/java/dev/number6/keyphrases/dagger/FakeComprehensionResultsModule.kt
package dev.number6.keyphrases.dagger

import dagger.Module
import dagger.Provides
import dev.number6.comprehend.results.PresentableKeyPhrasesResults
import dev.number6.message.ChannelMessages
import dev.number6.message.ChannelMessagesToComprehensionResultsFunction
import dev.number6.message.ComprehensionResultsConsumer
import org.junit.jupiter.api.Assertions
import java.util.*
import javax.inject.Singleton

@Module
class FakeComprehensionResultsModule {
    @Provides
    @Singleton
    fun providesMessageToEntityResults(): ChannelMessagesToComprehensionResultsFunction<PresentableKeyPhrasesResults> {
        return ConfigurableResultsFunction()
    }

    @Provides
    @Singleton
    fun providesEntityResultsConsumer(): ComprehensionResultsConsumer<PresentableKeyPhrasesResults> {
        return RecordingComprehensionResultsConsumer()
    }

    class ConfigurableResultsFunction : ChannelMessagesToComprehensionResultsFunction<PresentableKeyPhrasesResults> {
        private var presentableKeyPhrasesResults: PresentableKeyPhrasesResults? = null
        override fun apply(channelMessages: ChannelMessages): PresentableKeyPhrasesResults {
            return presentableKeyPhrasesResults
                    ?: Assertions.fail("presentable key phrases results needs to be configured")
        }

        fun setPresentableKeyPhrasesResults(presentableKeyPhrasesResults: PresentableKeyPhrasesResults?) {
            this.presentableKeyPhrasesResults = presentableKeyPhrasesResults
        }
    }

    class RecordingComprehensionResultsConsumer : ComprehensionResultsConsumer<PresentableKeyPhrasesResults> {
        private val resultsConsumed: MutableList<PresentableKeyPhrasesResults> = ArrayList()
        fun getResultsConsumed(): MutableList<PresentableKeyPhrasesResults> {
            return resultsConsumed
        }

        override fun accept(presentableKeyPhrasesResults: PresentableKeyPhrasesResults) {
            resultsConsumed.add(presentableKeyPhrasesResults)
        }
    }
}